src/infrastructure/gateway/RegistrationInMemoryGateway.ts

```diff
@@ -0,0 +1,110 @@
+/* eslint-disable */
+
+import crypto from "crypto";
+import {
+  LimitedPartnership,
+  NameEndingType,
+} from "@companieshouse/api-sdk-node/dist/services/limited-partnerships";
+
+import PageRegistrationType from "../../presentation/controller/registration/PageRegistrationType";
+import IRegistrationGateway from "../../domain/IRegistrationGateway";
+import CustomError from "../../domain/entities/CustomError";
+import TransactionLimitedPartnership from "../../domain/entities/TransactionLimitedPartnership";
+import LimitedPartnershipBuilder from "./LimitedPartnershipBuilder";
+
+class RegistrationInMemoryGateway implements IRegistrationGateway {
+  transactionId = crypto.randomUUID().toString();
+  submissionId = crypto.randomUUID().toString();
+
+  limitedPartnerships: TransactionLimitedPartnership[] = [];
+  errors: CustomError[] = [];
+
+  feedLimitedPartnerships(
+    limitedPartnerships: TransactionLimitedPartnership[]
+  ) {
+    this.limitedPartnerships = limitedPartnerships;
+  }
+
+  feedErrors(errors: CustomError[]) {
+    this.errors = errors;
+  }
+
+  async getSubmissionById(id: string): Promise<LimitedPartnership> {
+    const limitedPartnerShip = this.limitedPartnerships.find(
+      (lp) => lp._id === id
+    );
+
+    if (!limitedPartnerShip) {
+      throw new CustomError("Limited partnership", `Not found: ${id}`);
+    }
+
+    return new LimitedPartnershipBuilder(limitedPartnerShip).build();
+  }
+
+  async createTransaction(pageType: PageRegistrationType): Promise<string> {
+    if (pageType === PageRegistrationType.name) {
+      return this.transactionId;
+    }
+
+    return "";
+  }
+
+  async createTransactionAndFirstSubmission(
+    pageType: PageRegistrationType,
+    transactionId: string,
+    data: Record<string, any>
+  ): Promise<string> {
+    if (!data.partnership_name) {
+      this.errors.push(
+        new CustomError(
+          "partnership_name",
+          "partnership_name must be at least 3 characters"
+        )
+      );
+    }
+
+    if (!data.name_ending) {
+      this.errors.push(
+        new CustomError(
+          "name_ending",
+          `name_ending must be one of ${Object.keys(NameEndingType).join(", ")}`
+        )
+      );
+    }
+
+    if (this.errors.length > 0) {
+      throw this.errors;
+    }
+
+    const limitedPartnership = this.buildLimitedPartnership(
+      pageType,
+      data,
+      transactionId
+    );
+
+    this.limitedPartnerships.push(limitedPartnership);
+
+    if (transactionId === this.transactionId) {
+      this.limitedPartnerships[0] = limitedPartnership;
+    }
+
+    return this.submissionId;
+  }
+
+  private buildLimitedPartnership(
+    pageType: PageRegistrationType,
```

Comment:
I think this builder code is something new added to the PR? Not sure why we need it so will need a chat.

Reply:
the builder was already there
it gets the data sent and inserts it in the correct place in LimitedPartnership
because the api is waiting for the full object not just the current data

Reply:
👍 not sure what you mean by "correct place" though. `pageType` doesn't seem to be used by the builder method called here.

Reply:
for example, the data sent by the address page will probably be inserted into a sub-object called office_address or something like that
we will use type to insert the received data in the sub-object and not in the root ( data: { ... } )

Reply:
👍 makes sense now.